xds/internal/balancer/wrrlocality/balancer.go

```diff
@@ -45,24 +49,155 @@ func (bb) Name() string {
+func (a AddrInfo) Equal(o interface{}) bool {
+	oa, ok := o.(AddrInfo)
+	return ok && oa.LocalityWeight == a.LocalityWeight
+}
+
+// AddrInfo is the locality weight of the locality an address is a part of.
+type AddrInfo struct {
+	LocalityWeight uint32
+}
+
+// SetAddrInfo returns a copy of addr in which the BalancerAttributes field is
+// updated with AddrInfo.
+func SetAddrInfo(addr resolver.Address, addrInfo AddrInfo) resolver.Address {
+	addr.BalancerAttributes = addr.BalancerAttributes.WithValue(attributeKey{}, addrInfo)
+	return addr
+}
+
+func (a AddrInfo) String() string {
+	return fmt.Sprintf("Locality Weight: %d", a.LocalityWeight)
+}
+
+// getAddrInfo returns the AddrInfo stored in the BalancerAttributes field of
+// addr. Returns false if no AddrInfo found.
+func getAddrInfo(addr resolver.Address) (AddrInfo, bool) {
+	v := addr.BalancerAttributes.Value(attributeKey{})
+	ai, ok := v.(AddrInfo)
+	return ai, ok
+}
+
+// wrrLocalityBalancer wraps a weighted target balancer, and builds
+// configuration for the weighted target once it receives configuration
+// specifying the weighted target child balancer and locality weight
+// information.
+type wrrLocalityBalancer struct {
+	// child will be a weighted target balancer, and this balancer will build
+	// configuration for this child. Thus, build it at wrrLocalityBalancer build
+	// time, and configure it once wrrLocalityBalancer received configurations.
+	// Other balancer operations you pass through.
+	child balancer.Balancer
+
+	configParser balancer.ConfigParser
+
+	logger *grpclog.PrefixLogger
+}
+
+func (b *wrrLocalityBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		b.logger.Errorf("received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	weightedTargets := make(map[string]weightedtarget.Target)
+	for _, addr := range s.ResolverState.Addresses {
+		// This get of LocalityID could potentially return a zero value. This
+		// shouldn't happen though (this attribute that is set actually gets
+		// used to build localities in the first place), and thus don't error
+		// out, and just build a weighted target with undefined behavior.
+		locality := internal.GetLocalityID(addr)
+		localityString, err := locality.ToString()
+		if err != nil {
+			// Should never happen.
+			logger.Infof("failed to marshal LocalityID: %v, skipping this locality in weighted target")
+		}
+		ai, ok := getAddrInfo(addr)
+		if !ok {
+			return fmt.Errorf("addr: %v is misisng locality weight information", addr)
+		}
+		weightedTargets[localityString] = weightedtarget.Target{Weight: ai.LocalityWeight, ChildPolicy: lbCfg.ChildPolicy}
+	}
+	wtCfg := &weightedtarget.LBConfig{Targets: weightedTargets}
+	wtCfgJSON, err := json.Marshal(wtCfg)
+	if err != nil {
+		// Shouldn't happen.
+		return fmt.Errorf("error marshalling prepared wtCfg: %v", wtCfg)
```

Comment:
Same here. Prefer
`return fmt.Errorf("xds: failed to marshal weighted target configuration: %v", wtCfg)`

Reply:
Discussed offline; will switch component name to balancer name

Reply:
I still don't see the change here.

Reply:
As above, I added prefix xds_wrr_locality.